src/main/scala/com/microsoft/hyperspace/index/IndexLogEntry.scala

```diff
@@ -29,19 +35,180 @@ case class NoOpFingerprint() {
+      throwIfNotExists: Boolean = false): Content =
+    Content(Directory.fromDirectory(path, pathFilter, throwIfNotExists))
+
+  /**
+   * Create a Content object from a specified list of leaf files. Any files not listed here will
+   * NOT be part of the returned object.
+   *
+   * @param files List of leaf files.
+   * @return Content object with Directory tree from leaf files.
+   */
+  def fromLeafFiles(files: Seq[FileStatus]): Content = Content(Directory.fromLeafFiles(files))
+}
+
+case class Directory(name: String, files: Seq[FileInfo] = Seq(), subDirs: Seq[Directory] = Seq())
+
+object Directory {
+
+  /**
+   * Create a Directory object from a directory path by recursively listing its leaf files. All
+   * files from the directory tree will be part of the Directory.
+   *
+   * @param path Starting directory path under which the files will be considered part of the
+   *             Directory object.
+   * @param pathFilter Filter for accepting paths. The default filter is picked from spark
+   *                   codebase, which filters out files like _SUCCESS.
+   * @param throwIfNotExists If true, throw FileNotFoundException if path is not found. If set to
+   *                         false, create a blank directory tree with no files.
+   * @return Directory tree starting at root, and containing the files from "path" argument.
+   */
+  def fromDirectory(
+      path: Path,
+      pathFilter: PathFilter = PathUtils.DataPathFilter,
+      throwIfNotExists: Boolean = false): Directory = {
+    val fs = path.getFileSystem(new Configuration)
+    val leafFiles = listLeafFiles(path, pathFilter, throwIfNotExists, fs)
+
+    if (leafFiles.nonEmpty) {
+      fromLeafFiles(leafFiles)
+    } else {
+      val rootPath = getRoot(path)
+      if (rootPath.equals(path)) {
+        createDirectory(path, Array())
+      } else {
+        Directory(rootPath.toString, Seq(), Seq(createDirectory(path, Array())))
+      }
+    }
+  }
+
+  /**
+   * Create a Content object from a specified list of leaf files. Any files not listed here will
+   * NOT be part of the returned object
+   * Pre-requisite: files list should be non-empty.
+   * Pre-requisite: all files must be leaf files.
+   *
+   * @param files List of leaf files.
+   * @return Content object with Directory tree from leaf files.
+   */
+  def fromLeafFiles(files: Seq[FileStatus]): Directory = {
+    require(
+      files.nonEmpty,
+      s"Empty files list found while creating a ${Directory.getClass.getName}.")
+
+    require(
+      files.forall(!_.isDirectory),
+      "All files must be leaf files for creation of Directory.")
+
+    /* from org.apache.spark.sql.execution.datasources.InMemoryFileIndex. */
+    val leafDirToChildrenFiles = files.toArray.groupBy(_.getPath.getParent)
+
+    val subDirs = leafDirToChildrenFiles
+      .filterNot(_._1.isRoot)
+      .map {
+        case (dir, statuses) => createDirectory(dir, statuses)
```

Comment:
The limitation means: file list `["a/a/b/c.parquet", "a/a/c/d.parquet", "a/a/e.parquet"]` 
is stored as
```
root { name: "a", subDirs: [ { name: "a", subDirs: [ {name: "b", subDirs[], files: ["c.parquet"]}] }, 
                             { name: "a", subDirs: [ {name: "c", subDirs[], files: ["d.parquet"]}] }]
                             { name: "a", subDirs: [], files["e.parquet"] } ] }
```

I think it is more like a list and should be fixed.
```
[(dir:"a/a/b", files:["c.parquet"]), 
 (dir:"a/a/c", files["d.parquet"]),
 (dir:"a/a", files:["e.parquet"])]
```

cc: @imback82

Reply:
Sorry, I didn't fully get your point. Are you suggesting to store this in a different format?

Reply:
The structure should be like the following:
```
root { name: "a", subDirs: [ { name: "a", subDirs: [ {name: "b", subDirs[], files: ["c.parquet"]}, 
                                                     {name: "c", subDirs[], files: ["d.parquet"]}],
                               files["e.parquet"] } ] }
```

Reply:
Got it. @apoorvedave1 I think this sounds more like a bug than a limitation since we shouldn't have sub-directories with the same name.

Reply:
@imback82 I think there is some confusion. According to @sezruby , the first and second directories in the tree have the same name which leads to root and subdir both having name "a". 
The original file list itself is '["a/a/b/c.parquet", "a/a/c/d.parquet", "a/a/e.parquet"]'

Reply:
thanks @imback82, @sezruby . I think we went with this decision at the beginning of the pr right? I guess that's why I never thought it as a bug.
Please let me know if you want this to be fixed in this PR itself?

Reply:
OK, if so, I misunderstood the decision in the beginning. You can do it as a follow up, but let's do it right away.

Reply:
thanks @imback82 , @sezruby , after offline discussions it was decided to fix this issue in this PR itself. Please have a look at the updated code. I will mark the issue as resolved, please feel free to reopen if reqd